rfcs/20210305-tfx-struct2tensor.md

```diff
@@ -0,0 +1,221 @@
+is a library that helps build such ranking models. Supporting
+`tensorflow_ranking` in TFX has been a hot feature request.
+
+<div align="center">
+<img src='20210305-tfx-struct2tensor/tf_example_vs_elwc.png', width='700'>
+<p><i>
+  left: flat data represented by tf.Examples<br>
+  right: typical data for ranking problems -- each “example” contains
+  several “candidates”
+</i></p>
+</div>
+
+While it’s possible to encode anything in `tf.Examples`, this approach poses
+challenges to any component that needs to understand the data (e.g. Data
+Validation and Model Validation), and would also lead to bad user experience as
+they are forced to devise hacks.
+
+It’s also possible to address the problem in a case-by-case fashion by making
+TFX support a standard “container format” for each category of problem. We have
+compared that with the generic solution based on `struct2tensor` in previous
+efforts and concluded that we do
+not want another first-class citizen container format.
+
+Given that `struct2tensor` is able to decode an arbitrary protobuf (thus a good
+subset of all kinds of structured data) into a Tensor representation that
+preserves the structure (`tf.RaggedTensor`), we propose to
+solve the problem of supporting structured data in TFX through supporting
+`struct2tensor`.
+
+Thanks to Standardized TFX Inputs, a large portion of the solution is to create
+a `TFXIO` implementation for `struct2tensor`, and (as we will see later), the
+proper orchestration support needed for instantiating such a `TFXIO` in
+components.
+
+## Design Proposal
+
+### `GraphToTensorTFXIO`
+
+<div align="center"><img src='20210305-tfx-struct2tensor/graph_to_tensor_tfxio.png', width='700'></div>
+
+The diagram above shows how the proposed `GraphToTensorTFXIO` works:
+
+*   (1) The “Proto storage” is a format that Apache Beam can read from and
+    produce `PCollection[bytes]`. While the most naive example of such a format
+    is TFRecord, it does not have to be a row-based format. The only requirement
+    is that Beam can read it and produce `PCollection[bytes]`.
+
+*   (2) It relies on the fact that the `struct2tensor` query can be compiled to
+    a TF graph that converts a string tensor (containing serialized protos) to a
+    bunch of composite tensors, and thus can be stored in a file (SavedModel).
+
+*   (3) For beam-based components, `TFXIO` creates a PTransform that: decodes
+    the serialized records of protos to (batched) tensors using the saved TF
+    graph converts the tensors to arrow RecordBatches.
+
+*   (4) `TFXIO` will also create `TensorRepresentations` according to the output
+    signature of the saved TF graph, so that the following is identity: PICTURE
+    3
+
+*   (5) For TF trainers, `TFXIO` creates a `tf.data.Dataset` that:
+
+    -   reads the serialized records of protobufs as a string tensor
+    -   `.map()` the string tensor to decoded it into tensors using the saved
+        `struct2tensor` query.
```

Comment:
One question we have here is: How are you supposed to use structured datasets with tf.transform. Afaik TF.Transform outputs tf.example which is fine since on trainer then it will load fine, but you will need to access the dataview on trainer to create your serving signature.

Not sure if I'm missing something here.

CC @newtonle

Reply:
There are two ways:

1. You can disable TF Transform to output transformed examples [1]. Then, TF Transform will only output a TF Graph that contains the preprocessing_fn. That preprocessing_fn can already take tf.RaggedTensors and output tf.RaggedTensors, which struct2tensor can produce.

1. Potentially TF Transform can allow encoding tf.RaggedTensors and arbitrary tf.SparseTensors into tf.Examples. There are already some related work going on there.

[1] https://github.com/tensorflow/tfx/blob/c35e625ae058c35f80f9f71918071b840a8e787c/tfx/examples/ranking/ranking_pipeline.py#L105

Reply:
> but you will need to access the dataview on trainer to create your serving signature.

Trainer will offer an API to access the DataView (in the form of a tf.function). Currently though, since that tf.function essentially lives in the module file where the trainer code lives, you can get it without getting the DataView (https://github.com/tensorflow/tfx/blob/c35e625ae058c35f80f9f71918071b840a8e787c/tfx/examples/ranking/ranking_utils.py#L96)

Reply:
Would be interesting to learn on 2.

Regarding 1, would be nice to see an example where TFT tries to manipulate the ragged tensors. The Ranking example seems to not do anything (just calculate a vocabulary).

I think having it would be nice. Since relying on being in the same module may have unintended side effects (like if I'm using a newer trainer run_fn for an old dataview).

Reply:
> Regarding 1, would be nice to see an example where TFT tries to manipulate the ragged tensors. The Ranking example seems to not do anything (just calculate a vocabulary).
you can do all things supported by TF / TFT regarding those ragged tensors provided as the input to the preprocessing_fn. Note that TFT may not have a lot of analyzers that support RaggedTensors, but on the other hand, those analyzers may not need to understand the RaggedTensor's structure, which means you can simply apply the analyers against the RaggedTensor's `.flat_value`, then construct another RaggedTensor which has exactly the same structure, but with flat values replaced by the transformed values.


> I think having it be provided in the API (guessing TrainerArgs) would be nice. Since relying on being in the same module may have unintended side effects (like if I'm using a newer trainer run_fn for an old dataview).

Yes. That API will come.

Reply:
>> I think having it be provided in the API (guessing TrainerArgs) would be nice. Since relying on being in the same module may have unintended side effects (like if I'm using a newer trainer run_fn for an old dataview).

> Yes. That API will come.

https://github.com/tensorflow/tfx/commit/d2052ecddb47ac8f1a6162254b91427cbcc9ba8b